File: GO/DoublyLinkedList.go
package main

import "fmt"

type DLLNode struct {
	data string
	prev *DLLNode
	next *DLLNode
}

type DoublyLinkedList struct {
	head *DLLNode
	tail *DLLNode
	size int
}

func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{}
}

func (d *DoublyLinkedList) PushBack(value string) {
	newNode := &DLLNode{data: value}
	if d.head == nil {
		d.head = newNode
		d.tail = newNode
	} else {
		d.tail.next = newNode
		newNode.prev = d.tail
		d.tail = newNode
	}
	d.size++
}

func (d *DoublyLinkedList) PushFront(value string) {
	newNode := &DLLNode{data: value}
	if d.head == nil {
		d.head = newNode
		d.tail = newNode
	} else {
		newNode.next = d.head
		d.head.prev = newNode
		d.head = newNode
	}
	d.size++
}

func (d *DoublyLinkedList) PopFront() {
	if d.head == nil {
		return
	}
	d.head = d.head.next
	if d.head != nil {
		d.head.prev = nil
	} else {
		d.tail = nil
	}
	d.size--
}

func (d *DoublyLinkedList) PopBack() {
	if d.tail == nil {
		return
	}
	d.tail = d.tail.prev
	if d.tail != nil {
		d.tail.next = nil
	} else {
		d.head = nil
	}
	d.size--
}

func (d *DoublyLinkedList) Insert(index int, value string) {
	if index > d.size {
		return
	}
	if index == 0 {
		d.PushFront(value)
		return
	}
	if index == d.size {
		d.PushBack(value)
		return
	}
	
	current := d.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	
	newNode := &DLLNode{data: value}
	newNode.prev = current.prev
	newNode.next = current
	current.prev.next = newNode
	current.prev = newNode
	d.size++
}

func (d *DoublyLinkedList) Remove(index int) {
	if index >= d.size {
		return
	}
	if index == 0 {
		d.PopFront()
		return
	}
	if index == d.size-1 {
		d.PopBack()
		return
	}
	
	current := d.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	
	current.prev.next = current.next
	current.next.prev = current.prev
	d.size--
}

func (d *DoublyLinkedList) Get(index int) (string, error) {
	if index >= d.size {
		return "", fmt.Errorf("index out of range")
	}
	
	current := d.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	return current.data, nil
}

func (d *DoublyLinkedList) GetFront() (string, error) {
	if d.head == nil {
		return "", fmt.Errorf("list is empty")
	}
	return d.head.data, nil
}

func (d *DoublyLinkedList) GetBack() (string, error) {
	if d.tail == nil {
		return "", fmt.Errorf("list is empty")
	}
	return d.tail.data, nil
}

func (d *DoublyLinkedList) IsEmpty() bool {
	return d.size == 0
}

func (d *DoublyLinkedList) Size() int {
	return d.size
}

func (d *DoublyLinkedList) Clear() {
	for !d.IsEmpty() {
		d.PopFront()
	}
}

func (d *DoublyLinkedList) ToVector() []string {
	result := make([]string, 0, d.size)
	current := d.head
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	return result
}

func (d *DoublyLinkedList) FromVector(elements []string) {
	d.Clear()
	for _, element := range elements {
		d.PushBack(element)
	}
}

File: C++/doubly_linked_list.cpp
#include "doubly_linked_list.h"

DoublyLinkedList::DoublyLinkedList() : head(nullptr), tail(nullptr), list_size(0) {}

DoublyLinkedList::~DoublyLinkedList() {
    clear();
}

void DoublyLinkedList::push_back(const string& value) {
    Node* new_node = new Node(value);
    if (!head) {
        head = tail = new_node;
    } else {
        tail->next = new_node;
        new_node->prev = tail;
        tail = new_node;
    }
    list_size++;
}

void DoublyLinkedList::push_front(const string& value) {
    Node* new_node = new Node(value);
    if (!head) {
        head = tail = new_node;
    } else {
        new_node->next = head;
        head->prev = new_node;
        head = new_node;
    }
    list_size++;
}

void DoublyLinkedList::pop_front() {
    if (!head) return;
    
    Node* temp = head;
    head = head->next;
    if (head) {
        head->prev = nullptr;
    } else {
        tail = nullptr;
    }
    delete temp;
    list_size--;
}

void DoublyLinkedList::pop_back() {
    if (!tail) return;
    
    Node* temp = tail;
    tail = tail->prev;
    if (tail) {
        tail->next = nullptr;
    } else {
        head = nullptr;
    }
    delete temp;
    list_size--;
}

void DoublyLinkedList::insert(size_t index, const string& value) {
    if (index > list_size) return;
    
    if (index == 0) {
        push_front(value);
        return;
    }
    if (index == list_size) {
        push_back(value);
        return;
    }

    
    Node* current = head;
    for (size_t i = 0; i < index - 1; ++i) { 
        current = current->next;
    }
    
    
    Node* new_node = new Node(value);
    new_node->prev = current;
    new_node->next = current->next;
    
    if (current->next) {
        current->next->prev = new_node;
    }
    current->next = new_node;
    
    // Если вставляем в конец, обновляем tail
    if (new_node->next == nullptr) {
        tail = new_node;
    }
    
    list_size++;
}

void DoublyLinkedList::remove(size_t index) {
    if (index >= list_size) return;
    
    if (index == 0) {
        pop_front();
        return;
    }
    if (index == list_size - 1) {
        pop_back();
        return;
    }

    // Находим элемент для удаления
    Node* current = head;
    for (size_t i = 0; i < index; ++i) {
        current = current->next;
    }
    
    // Удаляем current
    if (current->prev) {
        current->prev->next = current->next;
    }
    if (current->next) {
        current->next->prev = current->prev;
    }
    
    delete current;
    list_size--;
}

string& DoublyLinkedList::get(size_t index) {
    if (index >= list_size) throw out_of_range("Index out of range");
    
    Node* current = head;
    for (size_t i = 0; i < index; ++i) {
        current = current->next;
    }
    return current->data;
}

string& DoublyLinkedList::get_front() {
    if (!head) throw out_of_range("List is empty");
    return head->data;
}

string& DoublyLinkedList::get_back() {
    if (!tail) throw out_of_range("List is empty");
    return tail->data;
}

bool DoublyLinkedList::isEmpty() const {
    return list_size == 0;
}

size_t DoublyLinkedList::size() const {
    return list_size;
}

void DoublyLinkedList::clear() {
    while (!isEmpty()) {
        pop_front();
    }
}

vector<string> DoublyLinkedList::toVector() const {
    vector<string> result;
    Node* current = head;
    while (current) {
        result.push_back(current->data);
        current = current->next;
    }
    return result;
}

void DoublyLinkedList::fromVector(const vector<string>& elements) {
    clear();
    for (const auto& element : elements) {
        push_back(element);
    }
}